pkg/clusterconditions/cache/cache.go

```diff
@@ -0,0 +1,169 @@
+// Package cache implements a throttled, caching condition.
+package cache
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/cluster-version-operator/pkg/clusterconditions"
+)
+
+// MatchResult represents a single Match invocation.
+type MatchResult struct {
+	// When is the completion time of the wrapped Match call.
+	When time.Time
+
+	// Access is the time of the most recent access.
+	Access time.Time
+
+	// Match is whether the result was a match or not.
+	Match bool
+
+	// Error is the error returned by the Match call.
+	Error error
+}
+
+// Cache wraps a cluster condition with caching and throttling.
+type Cache struct {
+	// Condition is the wrapped cluster condition.
+	Condition clusterconditions.Condition
+
+	// LastMatch is the completion time of the most recent Match
+	// call evaluated by the wrapped cluster condition.
+	LastMatch time.Time
+
+	// MinBetweenMatches is the minimum duration before a new
+	// Match call may be evaluated by the wrapped cluster condition.
+	MinBetweenMatches time.Duration
+
+	// MinForCondition is the minimum duration before a new Match
+	// call for a given condition may be evaluated.
+	MinForCondition time.Duration
+
+	// Expiration is the duration a value will be cached before
+	// being evicted for stale-ness.
+	Expiration time.Duration
+
+	// MatchResults holds results of previous match invocations.
+	MatchResults map[string]*MatchResult
+}
+
+// Valid returns an error if the wrapped cluster condition considers
+// the value invalid.
+func (c *Cache) Valid(ctx context.Context, condition *configv1.ClusterCondition) error {
+	return c.Condition.Valid(ctx, condition)
+}
+
+// Match returns the match value from the wrapped cluster condition,
+// possibly from a fresh evaluation, possibly from a local cache.
```

Comment:
Can you update this description to describe how it accomplishes this. And/or put some comments below on the major subsections describing what each is doing. I'm having a hard time following the overall flow.

Reply:
I've added more internal comments in 1ce08e79 -> a0b1136c.  I dunno if the details are relevant for callers, so I've left them out of the function's godocs for now, but feel free to re-open this thread or open a new one if you think we need more detail here.